localstack/aws/connect.py

```diff
@@ -115,6 +161,35 @@ def load_dto(data: str) -> InternalRequestParameters:
     return json.loads(data)
 
 
+T = TypeVar("T")
+
+
+class MetadataRequestInjector(Generic[T]):
+    def __init__(self, client: T):
+        self._client = client
+        self._params = None
+
+    def __getattr__(self, item):
+        target = getattr(self._client, item)
+        if not isinstance(target, Callable):
+            return target
+        if self._params:
+            return partial(target, **self._params)
+        else:
+            return target
+
+    def request_metadata(
+        self, source_arn: str | None = None, service_principal: str | None = None
+    ) -> Union[T, "MetadataRequestInjector[T]"]:
+        params = {}
+        if source_arn:
+            params["_SourceArn"] = source_arn
+        if service_principal:
+            params["_ServicePrincipal"] = service_principal
+        self._params = params
+        return self
```

Comment:
On second thought, I think it would be better to make this immutable and always return a new object